public/static/sdk/jsmr-0.1.js

```diff
@@ -145,22 +196,113 @@ var JsMr = Class.extend({
+         var state_params = {
+             Bucket: task.bucket_name,
+             Key: job_id.concat('_state'),
+         }
+
+         s3.getObject(state_params, function (err, data) {
+         if (err){
+	     console.log("Error in fetching state");
+             console.log(err, err.stack); // an error occurred
+	 }
+         else {
+             // successful response
+             this.current_state = JSON.parse( data.Body.toString());
+             context.state = this.current_state;
+	     console.log("Fetched state is: ");
+	     console.log(this.current_state);
+         }
+         });
+	
+        //Fetch data from AWS
+        var params = {
+            Bucket: task.bucket_name,
+            Key: task.object_key,
+            Range: 'bytes='.concat(task.start_index, '-', task.end_index)
+        };
+
+        var data_chunk;
+        s3.getObject(params, function (err, data) {
+            if (err){
+		console.log("Error in get Object");
+                console.log(err, err.stack); // an error occurred
+	    }
+            else {
+                // successful response
+		console.log("Successfull get object");
+                data_chunk = data.Body.toString();
+                var split_data = data_chunk.split("\n");
+
+                runMap().setup(context);
+                //run code on each line of data
+                for (var i = 0; i < split_data.length; i++) {
+                    var key = i + 1;
+                    var value = split_data[i];
+                    runMap().run(key, value, context);
+                }
+                _this.current_output = output;
+                _this.current_state = context.state;
+                //upload data to s3
+                _this.upload_data();
+            }
+        });
+
+    },
     /**
      * Run a new task, the client must be registered at this point. Also any
      * running tasks will be stopped.
      * @param  {Object} task The task info
      */
     runTask: function (task) {
+	console.log("In run task");
+        console.log(task);
+        var _this = this;
         if (task == null)
             return;
         if (!this.registered) {
-            this.log("Can't run Task `"+task.task_id+"`, client not registered", true);
-        }
-        if (this.current_task != null) {
-            // stop and cleanup the current task before starting a new one
+            this.log("Can't run Task `" + task.task_id + "`, client not registered", true);
         }
 
-        // TODO run the task !
+        // TODO In case of multiple tasks stop and cleanup the current task before starting a new one
+        //cleanup this.current_task
+        // Update current task with new one.
+        this.current_task = task;
+	
+/*        
+         if(this.step_list.indexof(this.current_task.step) == -1)
+         {
+             this.step_list.push(this.current_task.step);
+             this.getCode();
+         }
+*/         
+        this.getCode();
+	this.getData(task);
```

Comment:
Again, These two need to run after each other. currently you might get the code after the data which means you'll either end up not having `runMap()` or running the previous mapper on the data.